pkg/sql/crdb_internal.go

```diff
@@ -1632,39 +1638,45 @@ CREATE TABLE crdb_internal.table_indexes (
   is_inverted      BOOL NOT NULL
 )
 `,
-	populate: func(ctx context.Context, p *planner, dbContext *DatabaseDescriptor, addRow func(...tree.Datum) error) error {
+	generator: func(ctx context.Context, p *planner, dbContext *DatabaseDescriptor) (virtualTableGenerator, func(), error) {
 		primary := tree.NewDString("primary")
 		secondary := tree.NewDString("secondary")
-		return forEachTableDescAll(ctx, p, dbContext, hideVirtual,
-			func(db *DatabaseDescriptor, _ string, table *TableDescriptor) error {
-				tableID := tree.NewDInt(tree.DInt(table.ID))
-				tableName := tree.NewDString(table.Name)
-				if err := addRow(
-					tableID,
-					tableName,
-					tree.NewDInt(tree.DInt(table.PrimaryIndex.ID)),
-					tree.NewDString(table.PrimaryIndex.Name),
-					primary,
-					tree.MakeDBool(tree.DBool(table.PrimaryIndex.Unique)),
-					tree.MakeDBool(table.PrimaryIndex.Type == sqlbase.IndexDescriptor_INVERTED),
-				); err != nil {
-					return err
-				}
-				for _, idx := range table.Indexes {
+		worker := func(addRow func(...tree.Datum) error) error {
+			return forEachTableDescAll(ctx, p, dbContext, hideVirtual,
+				func(db *DatabaseDescriptor, _ string, table *TableDescriptor) error {
+					tableID := tree.NewDInt(tree.DInt(table.ID))
+					tableName := tree.NewDString(table.Name)
 					if err := addRow(
 						tableID,
 						tableName,
-						tree.NewDInt(tree.DInt(idx.ID)),
-						tree.NewDString(idx.Name),
-						secondary,
-						tree.MakeDBool(tree.DBool(idx.Unique)),
-						tree.MakeDBool(idx.Type == sqlbase.IndexDescriptor_INVERTED),
+						tree.NewDInt(tree.DInt(table.PrimaryIndex.ID)),
+						tree.NewDString(table.PrimaryIndex.Name),
+						primary,
+						tree.MakeDBool(tree.DBool(table.PrimaryIndex.Unique)),
+						tree.MakeDBool(table.PrimaryIndex.Type == sqlbase.IndexDescriptor_INVERTED),
 					); err != nil {
 						return err
 					}
-				}
-				return nil
-			})
+					for _, idx := range table.Indexes {
+						if err := addRow(
+							tableID,
+							tableName,
+							tree.NewDInt(tree.DInt(idx.ID)),
+							tree.NewDString(idx.Name),
+							secondary,
+							tree.MakeDBool(tree.DBool(idx.Unique)),
+							tree.MakeDBool(idx.Type == sqlbase.IndexDescriptor_INVERTED),
+						); err != nil {
+							return err
+						}
+					}
+					return nil
+				},
+			)
+		}
+		cleanup, start, next := setupGenerator(worker)
+		go start()
+		return next, cleanup, nil
```

Comment:
is there a reason we can't just return the `worker`, and let this boilerplate bit be somewhere else?

Reply:
There are some existing generators that don't necessarily fit into this pattern, so I don't want to push it up into the virtual table infrastructure yet.